File: JAVA/src/main/scala/engine/shader/Shader.scala
package engine.shader

/**
  * Created by Mnenmenth Alkaborin
  * Please refer to LICENSE file if included
  * for licensing information
  * https://github.com/Mnenmenth
  */

import engine.math.{Matrix4f, Vector3f}
import engine.util.Util
import org.lwjgl.opengl.GL20._
import org.lwjgl.opengl.GL32._

import scala.collection.mutable

class Shader {

  private val program: Int = glCreateProgram()
  private val uniforms = new mutable.HashMap[String, Int]()

  if(program == 0) println("Shader creation failed: Could not find valid memory location in constructor")

  def bind(): Unit ={
    glUseProgram(program)
  }

  def addUniform(uniform: String): Unit ={
    val uniformLocation = glGetUniformLocation(program, uniform)

    if(uniformLocation == -1) println("Error: Could not find uniform: " + uniform)

    uniforms.put(uniform, uniformLocation)
  }

  def addVertexShader(text: String): Unit ={
    addProgram(text, GL_VERTEX_SHADER)
  }

  def addFragmentShader(text: String): Unit ={
    addProgram(text, GL_FRAGMENT_SHADER)
  }

  def addGeoShader(text: String): Unit ={
    addProgram(text, GL_GEOMETRY_SHADER)
  }

  def compileShader(): Unit ={
    glLinkProgram(program)
    if(glGetProgrami(program, GL_LINK_STATUS) == 0) println(glGetShaderInfoLog(program, 1024))

    glValidateProgram(program)
    if(glGetProgrami(program, GL_VALIDATE_STATUS) == 0) println(glGetShaderInfoLog(program, 1024))
  }

  private def addProgram(text: String, pType: Int): Unit ={
    val shader = glCreateShader(pType)

    if(shader == 0) println("Shader creation failed: Could not find valid memory location when adding shader")

    glShaderSource(shader, text)
    glCompileShader(shader)

    if(glGetShaderi(shader, GL_COMPILE_STATUS) == 0) println(glGetShaderInfoLog(shader, 1024))

    glAttachShader(program, shader)

  }

  def setUniformi(uniformName: String, value: Int): Unit = glUniform1i(uniforms.getOrElse(uniformName,0), value)
  def setUniformf(uniformName: String, value: Float): Unit = glUniform1f(uniforms.getOrElse(uniformName,0), value)
  def setUniformv(uniformName: String, value: Vector3f): Unit = glUniform3f(uniforms.getOrElse(uniformName,0), value.x, value.y, value.z)
  def setUniformm(uniformName: String, value: Matrix4f): Unit = glUniformMatrix4fv(uniforms.getOrElse(uniformName,0), true, Util.createFlippedBuffer(value))

}

File: CPP/src/engine/shader/Shader.cpp
//
// Created by Razim on 3/2/2016.
//

#include <iostream>
#include "Shader.h"
#include "../util/Util.h"
#define GLEW_STATIC
#include "../../glew/glew.h"


Shader::Shader() : program(glCreateProgram()){ }

void Shader::bind() {
    glUseProgram(program);
}

void Shader::addUniform(const char *uniform) {
    GLint uniformLocation = glGetUniformLocation(program, uniform);
    if (uniformLocation == -1) std::cout << "Error: Could not find uniform: "  << uniform << std::endl;
    uniforms[uniform] = uniformLocation;
}

void Shader::addVertexShader(const char *source) {
    addProgram(source, GL_VERTEX_SHADER);
}

void Shader::addFragmentShader(const char *source) {
    addProgram(source, GL_FRAGMENT_SHADER);
}

void Shader::addGeoShader(const char *source) {
    addProgram(source, GL_GEOMETRY_SHADER);
}

void Shader::compileShader() {
    glLinkProgram(program);
    GLint link;
    glGetProgramiv(program, GL_LINK_STATUS, &link);

    if (link == 0) {
        GLchar infoLog;
        glGetShaderInfoLog(program, 1024, NULL, &infoLog);
        std::cout << infoLog << std::endl;
    }

    glValidateProgram(program);
    GLint validate;
    glGetProgramiv(program, GL_VALIDATE_STATUS, &validate);

    if (validate == 0) {
        GLchar infoLog;
        glGetShaderInfoLog(program, 1024, NULL, &infoLog);
        std::cout << infoLog << std::endl;
    }

}

void Shader::addProgram(const char *source, GLenum pType) {
    GLuint shader = glCreateShader(pType);
    if (shader == 0)
        std::cout << "Shader creation failed: Could not find valid memory location when adding shader" << std::endl;
    std::string length = source;
    GLint lengths[1];
    lengths[0] = length.length();
    glShaderSource(shader, 1, &source, lengths);
    glCompileShader(shader);
    GLint returns;
    glGetShaderiv(shader, GL_COMPILE_STATUS, &returns);
    if (returns == 0) {
        GLchar infoLog;
        glGetShaderInfoLog(shader, 1024, NULL, &infoLog);
    }
    glAttachShader(program, shader);
}

void Shader::setUniformi(std::string uniformName, GLint value) {
    glUniform1i(uniforms[uniformName], value);
}

void Shader::setUniformf(std::string uniformName, GLfloat value) {
    glUniform1f(uniforms[uniformName], value);
}

void Shader::setUniformv(std::string uniformName, Vector3f value) {
    glUniform3f(uniforms[uniformName], value.getX(), value.getY(), value.getZ());
}

void Shader::setUniformm(std::string uniformName, Matrix4f value) {
    printf("b\n");

    glUniform4fv(uniforms[uniformName], sizeof(value), Util::createFlippedBuffer(value));
    printf("e\n");

}
